explore/src/main/scala/explore/targeteditor/BrightnessesEditor.scala

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2016-2021 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package explore.targeteditor
+
+import cats.Order._
+import cats.syntax.all._
+import crystal.ViewF
+import crystal.react.View
+import crystal.react.implicits._
+import crystal.react.reuse._
+import eu.timepit.refined.auto._
+import explore.Icons
+import explore.components.ui.ExploreStyles
+import explore.implicits._
+import explore.model.display._
+import explore.utils.ReactTableHelpers
+import japgolly.scalajs.react._
+import japgolly.scalajs.react.callback.CallbackCats._
+import japgolly.scalajs.react.vdom.html_<^._
+import lucuma.core.enum.Band
+import lucuma.core.math.BrightnessUnits._
+import lucuma.core.math.BrightnessValue
+import lucuma.core.math.dimensional._
+import lucuma.core.util.Display
+import lucuma.core.util.Enumerated
+import lucuma.ui.forms.EnumViewSelect
+import lucuma.ui.optics.ChangeAuditor
+import lucuma.ui.optics.ValidFormatInput
+import lucuma.ui.reusability._
+import monocle.Focus
+import react.common._
+import react.semanticui.collections.form.Form
+import react.semanticui.collections.table._
+import react.semanticui.elements.button.Button
+import react.semanticui.elements.segment.Segment
+import react.semanticui.elements.segment.SegmentAttached
+import react.semanticui.sizes._
+import reactST.reactTable._
+import reactST.reactTable.mod.SortingRule
+
+import scala.collection.immutable.SortedMap
+
+sealed trait BrightnessesEditor[T] {
+  val brightnesses: View[SortedMap[Band, BrightnessMeasure[T]]]
+  val disabled: Boolean
+}
+
+sealed abstract class BrightnessesEditorBuilder[T, Props <: BrightnessesEditor[T]](implicit
+  enumUnits: Enumerated[Units Of Brightness[T]]
+) {
+  protected case class State(usedBands: Set[Band], newBand: Option[Band])
+
+  object State {
+    val usedBands = Focus[State](_.usedBands)
+    val newBand   = Focus[State](_.newBand)
+
+    def fromUsedBrightnesses(brightnesses: SortedMap[Band, BrightnessMeasure[T]]): State = {
+      val usedBands = brightnesses.keySet
+      State(usedBands, Band.all.diff(usedBands.toList).headOption)
+    }
+  }
+
+  protected def defaultBandUnits: Band => Units Of Brightness[T] // Abstract
+
+  implicit protected def propsReuse: Reusability[Props] // Abstract
+  implicit protected val stateReuse: Reusability[State] = Reusability.derive
+
+  implicit protected val displayTaggedUnits: Display[Units Of Brightness[T]] = Display[Units].narrow
```

Comment:
What is `narrow`?

Reply:
It adapts the instance of `Display[A]` as a `Display[B]`, where `B <: A`. I'll look into moving this to lucuma-core.